src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -487,6 +491,197 @@ test_budgets(void)
+	}
+}
+
+int wakeup_test_start = 0;
+/* this is preempted thread because, send with yield adds it to preempted */
+static void
+preempted_thd(void *d)
+{
+	struct exec_cluster *e = &(((struct activation_test_data *)d)->p);
+	struct exec_cluster *i = &(((struct activation_test_data *)d)->i);
+
+	while (1) {
+		PRINTC("\tPreempted: activating Intr\n");
+		if (wakeup_test_start) wakeup_test_start = 0;
+
+		cos_asnd(i->sc, 1);
+
+		if (!wakeup_test_start) {
+			PRINTC("\tPreempted: activated\n");
+			cos_switch(BOOT_CAPTBL_SELF_INITTHD_BASE, BOOT_CAPTBL_SELF_INITTCAP_BASE, 0, 0, 0, 0);
+		}
+	}
+}
+
+static void
+test_wakeup_case(struct activation_test_data *at, tcap_prio_t pprio, tcap_prio_t iprio, tcap_prio_t wprio)
+{
+	at->i.prio = iprio;
+	at->w.prio = wprio;
+	at->p.prio = pprio;
+	if (cos_tcap_transfer(at->p.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, at->p.prio)) assert(0);
+	if (cos_tcap_transfer(at->i.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, at->i.prio)) assert(0);
+	if (cos_tcap_transfer(at->w.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, at->w.prio)) assert(0);
+	wakeup_test_start = 1;
+	cos_switch(at->p.tc, at->p.tcc, at->p.prio, TCAP_TIME_NIL, 0, 0);
+}
+
+#define TEST_PRIO_HIGH (TCAP_PRIO_MAX)
+#define TEST_PRIO_MED  (TCAP_PRIO_MAX+1)
+#define TEST_PRIO_LOW  (TCAP_PRIO_MAX+2)
+
+static void
+test_wakeup(void)
+{
+	PRINTC("Testing Wakeup\n");
+	exec_cluster_alloc(&wat.s, intr_sched_thd, &wat, BOOT_CAPTBL_SELF_INITRCV_BASE);
+	exec_cluster_alloc(&wat.p, preempted_thd, &wat, BOOT_CAPTBL_SELF_INITRCV_BASE);
+	exec_cluster_alloc(&wat.i, intr_thd, &wat, wat.s.rc);
+	exec_cluster_alloc(&wat.w, worker_thd, &wat, wat.s.rc);
+	wat.s.prio = TEST_PRIO_HIGH; /* scheduler's prio doesn't matter */
+	if (cos_tcap_transfer(wat.s.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, wat.s.prio)) assert(0);
+
+	/*
+	 * test case 1: intr_thd = H, worker_thd = M, preempted_thd = L
+	 * expected result:
+	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
+	 * - cos_rcv from intr_thd should activate worker_thd.
+	 */
+	PRINTC(" Test - Intr = H, Worker = M, Preempted = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_LOW, TEST_PRIO_HIGH, TEST_PRIO_MED);
+
+	/*
+	 * test case 2: worker_thd = H, intr_thd = M, preempted_thd = L
+	 * expected result:
+	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
+	 * - cos_rcv from intr_thd should activate worker_thd.
+	 */
+	PRINTC(" Test - Worker = H, Intr = M, Preempted = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_LOW, TEST_PRIO_MED, TEST_PRIO_HIGH);
+
+	/*
+	 * test case 3: intr_thd = H, preempted_thd = M, worker_thd = L
+	 * expected result:
+	 * - cos_thd_wakeup should not invalidate preempted_thd.
+	 * - cos_rcv from intr_thd should activate preempted_thd.
+	 */
+	PRINTC(" Test - Intr = H, Preempted = M, Worker = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_MED, TEST_PRIO_HIGH, TEST_PRIO_LOW);
+
```

Comment:
Am I missing the logic that tests that the proper sequence of activations/switches actually happens?